### PR TITLE
Add documentation of some table column APIs in Storybook

### PR DESCRIPTION
**Pull request by jattasNI**
# Pull Request

## 🤨 Rationale

We've been adding new APIs to table columns in general and to specific column types like text but haven't been documenting them. This handles some of the tasks in #1044.

## 👩‍💻 Implementation

- add a new story dedicated to table columns. new story includes docs for:
    - column order
    - column header content
    - column sorting
    - other common attributes that don't merit their own story: field name and placeholder
- add a new story for the text column type
- remove some content from the existing table story
- Also did some minor refactoring to increase code sharing

There are additional stories tracked in #1044 that I'm not tackling yet like docs for action menu and column sizing.

## 🧪 Testing

Visual inspection of new stories

## ✅ Checklist

- [x] I have updated the project documentation to reflect my changes or determined no changes are needed.

### CODE REVIEW

**Review by atmgrifter00 — APPROVED** [review]

File: packages/nimble-components/src/table-column/tests/table-column.stories.ts
Comment (rajsite) [inline packages/nimble-components/src/table-column/tests/table-column.stories.ts:66]: What do you think about having this render under Table in the navigation tree? Kinda how Menu Item is under menu

I kinda like the organization of `Table/Column Configuration`:
image

File: packages/nimble-components/src/table-column/tests/table-column.stories.ts
Comment (rajsite) [inline packages/nimble-components/src/table-column/tests/table-column.stories.ts:1]: What about moving these files to table-column\base\tests? I think that's what these stories represent, the base features shared across all columns.

File: packages/nimble-components/src/table-column/tests/table-column.stories.ts
Comment (rajsite) [inline packages/nimble-components/src/table-column/tests/table-column.stories.ts:126]: Can you add some entry to the args table that shows something. Even a field that's like "Minimal Column Configuration Example" or something.

Right now from the Canvas view if I click it, I don't know what I'm supposed to take away from it:
image

File: packages/nimble-components/src/table-column/tests/table-column.stories.ts
Comment (rajsite) [inline packages/nimble-components/src/table-column/tests/table-column.stories.ts:163]: I think we should avoid having the usage warning render in the example code text. I'd argue the banner is enough in the rendered component and this just muddles trying to understand how it works. Maybe we can add the code-hide class to it?
image

**Review by rajsite — APPROVED** [review]

File: packages/nimble-components/src/table-column/tests/table-column.stories.ts
Comment (jattasNI) [reply packages/nimble-components/src/table-column/tests/table-column.stories.ts:66]: I like the idea but I can't figure out how to get it to work. If I use the `Table/Column Configuration` naming as described in [Storybook docs](https://storybook.js.org/docs/react/writing-stories/naming-components-and-hierarchy#grouping) I end up with a separate top-level section (sibling to Tokens and Tests). I did briefly see the purple folder layout in your screenshot but that only seems to appear after a hot reload on localhost; if I refresh the page after that I get the separate section.

image

I think a separate section for the table is worse than having multiple pages for column configuration so would vote to keep things as-is if we can't figure out how to get the layout in your screenshot to work reliably.

Comment (rajsite) [reply packages/nimble-components/src/table-column/tests/table-column.stories.ts]: I'm actually kinda feeling the table as a separate section. It's a sufficiently complex component with potentially dozen of interrelated elements to be aware of and lots of nuance.
But it's also a comparatively minor organization point and not worth blocking this PR. I'd rather the PR go in sooner so we can start capturing docs as features are merged and refine the organization more later.

File: packages/nimble-components/src/table-column/tests/table-column.stories.ts
Comment (jattasNI) [reply packages/nimble-components/src/table-column/tests/table-column.stories.ts]: I went back and forth on removing this story entirely since none of our other components that expose multiple stories do anything to address this. In the end I went with your suggestion but hope we can move away from the autogenerated docs tab and towards custom mdx docs someday. Then presumably we'd remove this placeholder story since I agree it is confusing on the Canvas tab.

**Comment by jattasNI** [comment]
@atmgrifter00 @rajsite while addressing review feedback I ended up adding a couple more stories: 
1. Table Column Configuration/Common Attributes to cover `column-hidden` and `column-id`
2. A story for `nimble-table-column-text` APIs `field-name` and `placeholder-text`

I'm resetting Meyer to get another set of eyes on the doc content. Milan is welcome to review too but didn't reset you (hoping the changes are uncontroversial so I can get this in quicker and save you review time).

File: packages/nimble-components/src/table-column/base/tests/table-column-stories-utils.ts
Comment (rajsite) [inline packages/nimble-components/src/table-column/base/tests/table-column-stories-utils.ts:22]: Can you follow a pattern similar to the `sharedMatrixParameters` pattern.
In-specific:
- have the resulting object be `as const`
- Don't use the shared value directly but "merge" it. ie the Args type should extend the CommonTableArgs type, spread the common table args into args, spread the common tableargstypes into argtypes, etc. This makes it easier for future story authors to see where to insert new args, etc.
- Be consistent with the naming convention, i.e. the shared prefix.

File: packages/nimble-components/src/table-column/base/tests/table-column-stories-utils.ts
Comment (rajsite) [inline packages/nimble-components/src/table-column/base/tests/table-column-stories-utils.ts:20]: `as const` and spread into usage spots

**Comment by rajsite** [comment]
> @atmgrifter00 @rajsite while addressing review feedback I ended up adding a couple more stories:
> 
> 1. Table Column Configuration/Common Attributes to cover `column-hidden` and `column-id`
> 2. A story for `nimble-table-column-text` APIs `field-name` and `placeholder-text`
> 
> I'm resetting Meyer to get another set of eyes on the doc content. Milan is welcome to review too but didn't reset you (hoping the changes are uncontroversial so I can get this in quicker and save you review time).

Just reviewed the code sharing pattern to match with existing patterns. Leaving review the actual doc content for the added stories to y'all.

**Review by rajsite — APPROVED** [review]

File: packages/nimble-components/src/table-column/base/tests/table-column.stories.ts
Comment (atmgrifter00) [inline packages/nimble-components/src/table-column/base/tests/table-column.stories.ts:207]:
```suggestion
const headerIconContent = 'If you provide icon content, you should set your own `title` on the icon element.';
```

**Pull request merged**
